daffodil-codegen-c/src/main/scala/org/apache/daffodil/codegen/c/generators/BinaryValueCodeGenerator.scala

```diff
@@ -93,11 +141,94 @@ trait BinaryValueCodeGenerator {
 
     val initERDStatement = ""
     val parseStatement =
-      s"""$indent1$indent2    parse_validate_fixed($field == $fixed, "$localName", pstate);
-         |$indent1$indent2    if (pstate->error) return;""".stripMargin
+      s"""$indent1$indent2    validate_fixed_attribute($field == $fixed, "$localName", &pstate->pu);
+         |$indent1$indent2    if (pstate->pu.error) return;""".stripMargin
+    val unparseStatement =
+      s"""$indent1$indent2    validate_fixed_attribute($field == $fixed, "$localName", &ustate->pu);
+         |$indent1$indent2    if (ustate->pu.error) return;""".stripMargin
+    cgState.addSimpleTypeStatements(initERDStatement, parseStatement, unparseStatement)
+  }
+
+  // Generate C code to validate a primitive element matches one of its enumeration values.
+  // May be replaced by a more specialized function in another trait for some types.
+  private def valueValidateEnumeration(
+    e: ElementBase,
+    deref: String,
+    enums: Array[String],
+    cgState: CodeGeneratorState,
+  ): Unit = {
+    val indent1 = if (cgState.hasChoice) INDENT else NO_INDENT
+    val indent2 = if (deref.nonEmpty) INDENT else NO_INDENT
+    val localName = cgState.cName(e)
+    val field = s"instance->$localName$deref"
+
+    // Extra initialization only for hexBinary enumerations
+    val arraysName = s"arrays_$localName"
+    val hexEnums = enums.zipWithIndex.map { case (s, index) =>
+      s"{$arraysName[$index], ${s.length / 2}, false}"
+    }
+    val hexEnumsInit = enums.map(_.grouped(2).map("0x" + _).mkString("{", ", ", "}"))
+    val enumsLenMax = enums.map(_.length / 2).max
+    val arraysInit =
+      s"""$indent1$indent2    uint8_t $arraysName[][$enumsLenMax] = ${hexEnumsInit.mkString(
+          "{",
+          ", ",
+          "}",
+        )};\n"""
```

Comment:
Are there any concerns that there might be a bunch of hex binary enums with one being much larger than the other, and so this array would have a lot of empty space? I guess the other option would be to create a bunch of separate variables (e.g. hexEnum1, hexEnum2, hexEnum3, ... 0), maybe that's not worth it. Or alternatively, concat the hex arrays into a single long array, and then then hexEnums have to correctly index into that array? Maybe that's too much complexity for a case that isn't that likely? I imagine most hex enums are all the same length, or at least close in length.

Reply:
I have not seen any schemas with hex binary enums yet, so let's put off any changes until we see hex enums actually used.  I only found out through experimentation that daffodil supports hex enums and thought daffodilC should try to support them too.  I think your imagination that most hex enums are close in length is probably right anyway.